File: LeetCode/040.combination-sum-ii/combination-sum-ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>>res = new ArrayList<>();
        Arrays.sort(candidates);
        solver(res, candidates, new ArrayList<>(), target, 0);
        return res;
    }

    public void solver(List<List<Integer>>res, int[] candidates, List<Integer>tmp, int target, int stIndex){
        if (target < 0) return;
        else if (target == 0)   res.add(new ArrayList<>(tmp));
        else
            for (int i = stIndex; i < candidates.length; i++){
                if (i > stIndex && candidates[i] == candidates[i - 1])  continue;
                if (target - candidates[i] < 0) break;
                tmp.add(candidates[i]);
                solver(res, candidates, tmp, target - candidates[i], i + 1);
                tmp.remove(tmp.size() - 1);
            }
    }
}

File: LeetCode/040.combination-sum-ii/combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>>res;
        vector<int>tmp;
        sort(candidates.begin(), candidates.end());
        solver(res, candidates, tmp, target, 0);
        return res;
    }

    void solver(vector<vector<int>>& res, vector<int>& candidates, vector<int>tmp, int target, int stIndex){
        if (target < 0) return;
        else if (target == 0)    res.push_back(tmp);
        else
            for (int i = stIndex; i < candidates.size(); i++){
                if (i > stIndex && candidates[i] == candidates[i - 1])    continue;
                tmp.push_back(candidates[i]);
                if (target - candidates[i] < 0) break;
                solver(res, candidates, tmp, target - candidates[i], i + 1);
                tmp.pop_back();
            }
    }
};